internal/db/embedding.go

```diff
@@ -0,0 +1,138 @@
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/philippgille/chromem-go"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/internal/keys"
+)
+
+func getEmbeddingFunc(provider, model, url string) (chromem.EmbeddingFunc, error) {
+	switch provider {
+	case "ollama":
+		return chromem.NewEmbeddingFuncOllama(model, url), nil
+	case "openai":
+		normalized := true
+		apiURL := url
+		if apiURL == "" {
+			apiURL = chromem.BaseURLOpenAI
+		}
+		return chromem.NewEmbeddingFuncOpenAICompat(
+			apiURL,
+			os.Getenv("OPENAI_API_KEY"),
+			model,
+			&normalized,
+		), nil
+	default:
+		return nil, NewErrUnknownEmbeddingProvider(provider)
+	}
+}
+
+// setEmbedding sets the embedding fields on the document if the related fields are dirty.
+// However, if the vector field itself has been set, it will not be overwritten by a new embedding generation.
+func (c *collection) setEmbedding(ctx context.Context, doc *client.Document, isCreate bool) error {
+	embeddingGenerated := false
+	for _, embedding := range c.Description().Embeddings {
+		vecValue, err := doc.GetValue(embedding.FieldName)
+		if err != nil && !errors.Is(err, client.ErrFieldNotExist) {
+			return NewErrGetEmbeddingField(err)
+		}
+		if vecValue != nil && vecValue.IsDirty() {
+			// If the vector has been explicitly set, no need to generate the embedding.
+			continue
+		}
+		fieldsVal := make(map[string]client.NormalValue)
+		needsGeneration := false
+		missingFieldsForGeneration := []client.FieldDefinition{}
+
+		// Get the new values of the fields used for embedding generation. We keep track
+		// of the fields that aren't defined to lookup their previous values later.
+		for _, embedField := range embedding.Fields {
+			if docField, ok := doc.Fields()[embedField]; ok {
+				if doc.Values()[docField].IsDirty() {
+					needsGeneration = true
+					fieldsVal[embedField] = doc.Values()[docField].NormalValue()
+				} else {
+					fieldDef, ok := c.def.GetFieldByName(embedField)
+					if !ok {
+						return NewErrEmbeddingFieldNotFound(embedField)
+					}
+					missingFieldsForGeneration = append(missingFieldsForGeneration, fieldDef)
+				}
+			}
+		}
+
+		if !needsGeneration {
+			continue
+		}
+
+		// If we are updating the document and we don't have all the fields used for vector embedding
+		// generation, we get the document to see if the fields have previously been set.
+		if !isCreate && len(missingFieldsForGeneration) > 0 {
+			oldDoc, err := c.get(
+				ctx,
+				keys.DataStoreKeyFromDocID(doc.ID()).ToPrimaryDataStoreKey(),
+				missingFieldsForGeneration,
+				false,
+			)
+			if err != nil {
+				return NewErrGetDocForEmbedding(err)
+			}
+			for _, embedField := range missingFieldsForGeneration {
+				if docField, ok := oldDoc.Fields()[embedField.Name]; ok {
+					fieldsVal[embedField.Name] = oldDoc.Values()[docField].NormalValue()
+				}
```

Comment:
suggestion: The whole block is showing as untested, please add tests if not there already (maybe due to ci failures, ignore if tested already).